kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaIterable.java

```diff
@@ -0,0 +1,450 @@
+      return createZipper(/*parents=*/ new ArrayList<>(), schema);
+    }
+
+    private static SchemaZipper createZipper(List<SchemaZipper> parents, DataType type) {
+      if (type instanceof ArrayType) {
+        ArrayType arrayType = (ArrayType) type;
+        return new ArraySchemaZipper(parents, arrayType);
+      } else if (type instanceof MapType) {
+        MapType mapType = (MapType) type;
+        return new MapSchemaZipper(parents, mapType);
+      } else if (type instanceof StructType) {
+        StructType structType = (StructType) type;
+        return new StructSchemaZipper(parents, structType);
+      } else {
+        throw new KernelException("Unsupported data type: " + type);
+      }
+    }
+
+    private static SchemaZipper createZipper(List<SchemaZipper> parents, StructField field) {
+      return createZipper(parents, field.getDataType());
+    }
+
+    /** Returns a zipper pointing the left-most child field of this zipper. */
+    public SchemaZipper childrenZipper() {
+      if (!hasChildren()) {
+        return null;
+      }
+      parents.add(this);
+      return createZipper(parents, fields.get(index));
+    }
+
+    /**
+     * Returns a zipper pointing to the next sibling of this zipper. (moving right across zippers).
+     */
+    public SchemaZipper moveToSibling() {
+      if (!hasMoreSiblings()) {
+        return null;
+      }
+      index++;
+      return this;
+    }
+
+    public boolean hasMoreSiblings() {
+      return index < fields.size() - 1;
+    }
+
+    @Override
+    public StructField getField() {
+      return currentField();
+    }
+
+    private SchemaZipper getParent() {
+      if (parents.isEmpty()) {
+        return null;
+      }
+      return parents.get(parents.size() - 1);
+    }
+
+    /**
+     * Returns a new zipper pointing to the parent of this zipper.
+     *
+     * <p>If the zipper has any modifications they are propagated up to the parent.
+     */
+    public SchemaZipper moveToParent() {
+      SchemaZipper parent = getParent();
+      if (!parents.isEmpty()) {
+        parents.remove(parents.size() - 1);
+      } else {
+        return null;
+      }
+      if (modified) {
+        // Propagate changes to parent.
+        StructField currentParentField = parent.currentField();
+        StructField newParentField = currentParentField.withDataType(extractDataTypeFromFields());
+        parent.updateField(newParentField);
+      }
+      return parent;
+    }
+
+    public DataType extractDataTypeFromFields() {
+      return constructType();
+    }
+
+    public StructField currentField() {
+      return fields.get(index);
+    }
+
+    @Override
+    public void updateField(StructField structField) {
+      try {
+        fields.set(index, structField);
+      } catch (UnsupportedOperationException e) {
+        // Field might be immutable, copy and set if this is the case.
+        fields = new ArrayList<>(fields);
+        fields.set(index, structField);
```

Comment:
why not just copy the fields to array and keep it locally? that is expected to be mutable and avoid the code like this.